core/platform/lf_rp2040_support.c

```diff
@@ -214,6 +216,170 @@ int _lf_single_threaded_notify_of_event() {
     sem_release(&_lf_sem_irq_event);
     return 0;
 }
+
+#else // LF_SINGLE_THREADED
+
+#warning "Threaded runtime on RP2040 is still experimental"
+
+/**
+ * @brief Get the number of cores on the host machine.
+ */
+int lf_available_cores() {
+    // Right now, runtime creates 1 main thread and 1 worker thread
+    // In the future, this may be changed to 2 (if main thread also functions
+    // as a worker thread)
+    return 1;
+}
+
+static void *(*thread_1) (void *);
+static void* thread_1_args;
+static int num_create_threads_called = 0;
+static semaphore_t thread_1_done;
+static void* thread_1_return;
+
+#define MAGIC_THREAD1_ID 314159
+
+void core1_entry() {
+    thread_1_return = thread_1(thread_1_args);
+    sem_release(&thread_1_done);
+
+    // infinite loop; core1 should never exit
+    while (1){
+        tight_loop_contents();
+    }
+}
+
+int lf_thread_create(lf_thread_t* thread, void *(*lf_thread) (void *), void* arguments) {
+    // make sure this fn is only called once
+    if (num_create_threads_called != 0) {
+        return 1;
+    }
+    thread_1 = lf_thread;
+    thread_1_args = arguments;
+    num_create_threads_called += 1;
+    sem_init(&thread_1_done, 0, 1);
+    multicore_launch_core1(core1_entry);
+    *thread = MAGIC_THREAD1_ID;
+    return 0;
+}
+
+int lf_thread_join(lf_thread_t thread, void** thread_return) {
+    if (thread != MAGIC_THREAD1_ID) {
+        return 1;
+    }
+    sem_acquire_blocking(&thread_1_done);
+    sem_release(&thread_1_done); // in case join is called again
+    if (thread_return) {
+        *thread_return = thread_1_return;
+    }
+    return 0;
+}
+
+int lf_mutex_init(lf_mutex_t* mutex) {
+    mutex_init(mutex);
+    return 0;
+}
+
+int lf_mutex_lock(lf_mutex_t* mutex) {
+    mutex_enter_blocking(mutex);
+    return 0;
+}
+
+int lf_mutex_unlock(lf_mutex_t* mutex) {
+    mutex_exit(mutex);
+    return 0;
+}
+
+int lf_cond_init(lf_cond_t* cond, lf_mutex_t* mutex) {
+    sem_init(&(cond->sema), 0, 1);
+    cond->mutex = mutex;
+    return 0;
+}
+
+int lf_cond_broadcast(lf_cond_t* cond) {
+    sem_reset(&(cond->sema), 1);
+    return 0;
+}
+
+int lf_cond_signal(lf_cond_t* cond) {
+    sem_reset(&(cond->sema), 1);
+    return 0;
+}
+
+int lf_cond_wait(lf_cond_t* cond) {
+    mutex_exit(cond->mutex);
+    sem_acquire_blocking(&(cond->sema));
+    mutex_enter_blocking(cond->mutex);
+    return 0;
+}
+
+int lf_cond_timedwait(lf_cond_t* cond, instant_t absolute_time_ns) {
+    absolute_time_t a = from_us_since_boot(absolute_time_ns / 1000);
+    bool acquired_permit = sem_acquire_block_until(&(cond->sema), a);
+    return acquired_permit ? 0 : LF_TIMEOUT;
+}
+
+
+// Atomics
```

Comment:
The atomics implementations are pretty much copy-pasted from Zephyr support. In a future PR, I would like to merge the two somehow. Not going to do it in this PR in case I end up accidentally breaking Zephyr support.

Reply:
Would be good to merge them yes. But another thought is that a scheduler using them extensively will be expensive on Zephyr/Pico and cheap on x86. Maybe we need a new scheduler for 32bit platforms without atomics? Or maybe not :)

Reply:
Definitely should test how much this affects performance once I start benchmarking on the rp2040. A new scheduler would be exciting :)